Handy vizualization tool:  http://www.pythontutor.com/visualize.html#mode=edit

# Week 3:  Acumulators and numbers

## Announcements

Some general reminders and points on things for this week:

* I will often put in extra requirements into the homework text in moodle, and these will be graded requirements. 
* Even a narrative only answer may still need you to run and deal with code, even if you aren't required to submit any code.
* Please read all the hints for each problem that I've given in the homework problems, and reference back to them as you have questions.  There are some where I have given you example outlines for your code solution.
* Please start to test your programs that take in inputs.  I've given you some reminders to test your code with many inputs, but I won't have that there the entire semester.

This week's chapter really focuses on numbers, both generating numbers, manipulating their contents, and converting them to suit your needs.  The only math we'll be doing is addition and division.  Nothing more.  Try to focus on understanding the code within the chapter and largely ignore the math that it's trying to talk about.  Along with numbers, the chapter introduces very important pattern that will be in many of your programs:  accumulators and counters.  This is where I'll be focusing my lecture on.  You will still be strongly needing the book with other content.   Any math that you need to know for your homework is explained in the hints area.  

As you are thinking of accumulators and counters, keep them separate in your head.  They're going to look really similar in the code, but their purpose and usage will be very different!  You can have multiple of both inside you program, but commonly you'll have an accumulator to collect up many values, and then a counter to keep track of how many things you have counter.  There is an argument to be made that counters are a special type of accumulator, but I prefer to think of them as siblings rather than parent/child.

We'll be talking about thinking through the application of many of these tools.  Not just how to generate the numbers that you want but how to think through solving problems with numbers and integrating those solutions into larger needs.

For each of these homework problems, you need to answer the following questions:

1. What kind of for loop do you need?
    * Are you repeating a certain number of times, and also don't care about the numbers generated by range()? (this would be a repetition loop)
    * Do you need to generate a series of numbers and collect those? (this would be an unpacking loop over range.
    * These two will not look much different, but will have one huge difference:  repetition loops will ignore the `i` value generated by that iconic pattern.  Unpacking loops to go over a sequence of numbers WILL use that number, and you'll need to fiddle with the arguments in `range()` to make it generate the numbers that you want.
2. Do you need an accumulator (summing up a series of values) or counter (just counting how many times something happens)?
3. Which value are you adding up for your accumulator?  Some of these problems have user input that's going into your accumulator.  Some of them will add up your iterable variable value (the thing produced by range in your for loop, if you're doing an unpacking loop with range being your sequence).

Also pay attention to two key points in this lecture or note them in the patterns as you read them later:

1. Where you put your accumulator or counter base is super important.  The bases need to come before the for loop taking the action of accumulating or counting.  Having your base inside that for loop means you'll erase all your results, and having it after you are trying to count will give you errors about the variable not existing.
2.  Where you put your print statements is also vitally important.  You aren't risking erasing your data here, but you are risking making strange outputs that are hiding what is going on.  Your print statements should be outside of your for loop (so after it, and indented to the left.)

Strongly consider using PythonTutor.com to develop your programs in, so you can see how all the values are changing.

# The two patterns:  accumulators vs counters

You'll note that the words much and many aren't specific here.  So when you are thinking about these questions, think about the complete question.

In particular, counting questions can be tricky here.  A counter literally has the name in there, but you may still want to use an accumulator.

The first question that you need to ask is:  what am I summing here?  Are you counting a series of uniform values?  Or are you counting up a series of different values?  The variability of the values is a big distinction here.

For example, we have a class of 24 students.  

**We can count the number of students.**  Each person here counts as "1".  This is a uniform value across all entities.  This is the core of a counter pattern.  Inside the loop you'll see that the incrementing value is often hard coded. 

**We can sum up the total age.** Each person has some age, which may repeat but is variable across everyone in here.  This value must be accessed each time through, and that value is added up to create a final value representing the total number of years old all the students are in this room.  

We can use them together.  If I know a total sum of a property for a set of entities, and the total number of entities, what does that sound like to you?

This is how we calculate the mean or average value.  So if we've got a loop counting these two things together, we can calculate some summary statistics about it.  

Can you think of some examples of things you might want to count up versus things you might want to sum up? 

So that's what we can do and how we will be mixing these up together.  Let's take a step back and look at some code for making these things happen.  


# Minimal counter

A minimal counter will be the easiest to see this mechanism in action.


## Counting up by a set number

For this task, we want to count how many times a loop executes.  This isn't a trick or anything fancy.  

Remember our steps to a for loop?  We need:

1. determine your pattern
2. identify your sequence
3. give your iterable a good name

Answering questions 1 and 2 here often go in circles.  Pick a pattern, think about how it'll work with the sequences available to you, and either rethink or continue forward.  As you build up more experience, you'll know what you need faster and begin to have a set of patterns that you are the most comfortable with.  As you are getting started, take note of the patterns that make the most sense to you and latch onto those.

Let's start with an range loop, because what we need is to make a for loop that can execute some number of times so that we can count how many times that is.  True, we can run a counter in every type of for loop out there, but let's go with a range loop to start with.

### Review a base range loop

``` python
for number in range(5):
    print("hello")
```

This block of code will print "hello" five times.  

We're going to start with this base, and change it so that we are counting how many times this runs.  We know it will be 5 times, but out end goal will be to have a variable holding this value.


### There are 4 steps to a counter

Here are the 4 steps:

1. Establish you base variable at the initial state (usually 0)
2. Start a for loop for whatever you're repeating
3. Do your stuff, whatever you need to do in that loop
4. Update your counter, usually by 1.  This is commonly the last thing in your loop.

The reference pattern is:

```python 
counter = 0 # 1: your counter variable at the initial state

for something in sequence: # 2: loop over something!
    do stuff # 3: do all the things
    counter = 1 + counter # 4: update the counter, replace 1 and/or the operator with whatever you need
```
 
My nitpick about the Zelle formula is that it leaves out the `do stuff` item that goes between your for loop and incrementing up your counter variable.  You are often having to do more than just update your counter variable, and many times you need that value in the process.  This means that you'll need to increment up your counter as the last thing you do in your loop. There will be exceptions to this, but they are rare.  More generally, all your incrementers should be at the end of your for loop, and the internal order doesn't always matter.

This particular example is counting up by 1, but you can always change what that number is or what the operator is.  For example, if you were counting up by 10, you would have 10 in step 4.


#### Worked example of a counter


##### Step 1: have a base

So let's start adapting things.  We need to start with a base.

``` python
count = 0
```

##### Step 2: have a for loop in there

Now we're going to add in our for loop.

``` python
count = 0

for number in range(5):
    print("hello")
```
You see that I haven't changed up the for loop's content, we'll get to that next.

##### Step 3: do stuff to get the count that you need in that for loop

Very often with counters you won't need to do anything here, because you're just brute force counting by a certain number each time.  But this is when you'd address any decision structures or data extraction methods to get out the value that you need to count by.

``` python
count = 0

for number in range(5):
    # you'd do stuff here
    # or maybe you have more counters do do
    # either way, all the fun stuff not counter related goes right after the for loop starts
    print("hello") # and likely print it out here instead of "hello"
```

##### Step 4:  increment up your counter

Now that we've got all our pieces in order and we've done our business, we can finally update our counter.  This is the last step because this is ofter the last thing your for loop.  One of the most common problems that I see is when the increment is put at the beginning of the for loop or in the middle of the action, and everything gets weird.  Almost always it will need to be the last line of code within your for loop.

``` python
count = 0

for number in range(5):
    count = count + 1 #here's the update
    
print(base)
```

Note our use of `base = base + 1` here.  We are updating a variable that already exists (this is why we define a base), and since we want to reuse it and have the value retained even when the for loop resets, we need to a) keep that base on the outside of our for loop and b) reuse the variable name.  So that repetition of the base variable name is purposeful and absolutely necessary.  

What would happen if we didn't include the `print()` there at the end?  (nothing would have printed out)

What would happen if the `print(base)` was indented inside the for loop? (we would see a print out for it each time).  Would we see count with the base value of 0?  Would we see count with the final value of 5?  Why?


# Minimal Accumulator:  summing variable/non-uniform values


Think like snow measurement, we say that there is an "accumulation of 3 inches".  You don't know how big the individual time moments are (like if it snowed heavily at once or if it was a slow trickle), but you do know the end total of how much snow ended up there.  You can also measure the amount incrementally over time, but once something is added in you can't know how big that individual piece is.

An accumulator is highlighted by the incremental value going in being variable in size.  Think of this like a running total where you are only keeping track of the current value, and once an individual value is added in, that value is lost.  

## There are 5 steps to an accumulator

Here are the 5 steps (they should look very familiar to you with the counter steps)

1. Establish you accumulator base at the initial state (usually 0)
2. Start a for loop for whatever you're repeating
3. Do your stuff, whatever you need to do in that loop
4. Determine your value to accumulate
5. Update your base variable by whatever your incremental value is.  This is commonly the last thing in your loop.

The reference pattern is:

```python 
total_something = 0 # 1: your counter variable at the initial state

for something in sequence: # 2: loop over something!
    do stuff # 3: do all the things
    value_you_want = ?? # 4: determine the value you want to add to your accumulator
    total_something = value + total_something # 5: update the counter, replace 1 and/or the operator with whatever you need
```

## Worked example

Let's sum up all the numbers produced in a range loop.  Yes, this is as silly example, but it keeps the clutter away so you can see the inner workings better.

Reminder, here's our range loop printing out 10 numbers, starting at 0.  We will be adding an accumulator pattern within and around this for loop.

``` python
for num in range(10):
    print(num)
```


### 1: add the base

We want to start at 0 because we're counting like humans.

``` python

total_sum = 0  # 1

for num in range(10):
    print(num)
```

### 2 & 3: looping and doing things

There's not much to do because we already have our loop set up for us.  Likewise, there's no transformation or lookup to do there.  

``` python

total_sum = 0 

for num in range(10): # 2
    # 3: you'd do stuff here to transform or extract the number that you need
    print(num) 
```

### 4: add the incrementing value

In this case, we're counting how many times the loop has executed.  

So how much do we want to count by each time?  1.

And where do we need to put the incrementing expression?  When in doubt, put it at the end and see what happens. In our case, we aren't doing much of anything, so we can put it anywhere.

Here's what the final form should look like.

In [2]:
total_sum = 0 # 1

for num in range(10):
    # we can get rid of the print now
    total_sum = num + total_sum # 5
    
print(total_sum)

45


# You can 'accumulate' more than numbers!!!!!

Anything that can 'hold' stuff can be accumulated into.  You'll be using this pattern a lot moving forward.

## Collecting letters

When a data type has the ability to be 'added' to, either through the `+` operator or a method (lists do this, but we aren't there yet), you can use this same kind of pattern.  Instead of setting the counter to `0`, you have an 'empty' whatever-it-is.  Since we've been playing with strings lately, let's use that as an example.

```python 
new_string = "" # 1 your empty string

for letter in "the quick brown fox jumped over the lazy dog": # 2 loop over something!
    do stuff # 3 maybe you're filtering for letter characters
    new_string = new_string + letter # 4 update the new string to include this letter
```

Other than not having numbers, you'll notice just one difference in this example:  I'm using `new_string = new_string + letter` whereas the counter had `thingtoadd = 1 + thingtoadd`.  So whereas numercial addition can have any order (Commutative Property) there will be times that order does matter.  In this case, it does with string concatenation.

`"fizzy" + "pop"` makes `"fizzypop"` and `"pop" + "fizzy"` makes `"popfizzy"`.  So the order that you decide to write the operation will determine what you get and should be part of your design process.  Most of the time you'll want to add new things on to the end, so:  `current + new`

## Generalizing the accumulator pattern beyond numbers

You should still keep your idea of the counter as something separate for now.  But we can come up with a more general pattern to serve you beyond basic operator (`+`) accumulation.

This pattern will be used all over as we move forward in class.  What will change is what the collection item is (a string, a sum, a file, list, etc), what you're adding to the thing, and what you need to do to create the thing you want to add to your collection.  Here's a more abstract formula:


```python 
collection_thing = ?? # declare an empty version of your collection thing 
# thing may also be opening an empty file to write out to

for thing in sequence: # loop over something!
    do stuff # transform, filter, etc
    ??? # put your new data in your collection thing
```

# common pitfalls

We're going to switch gears a bit and go over some of the common errors that you'll see related to these patterns.

## the accumulator value isn't adding or only has the last value

This happens when your counter variable is placed inside your for loop and ends up starting over each run. It will still perform the operations each time, but your counter will be reset each time through.  

In [1]:
for i in range(5):
    print("the number is:", i)
    total = 0
    total = total + i
    
print("the final result is:", total)

the number is: 0
the number is: 1
the number is: 2
the number is: 3
the number is: 4
the final result is: 4


All the pieces you need are in there, just in the wrong place.  I've fixed the code below by only moving the location of the `counter = 0` line.

In [2]:
total = 0 # look at me at the top now

for i in range(5):
    print("the number is:", i)
    total = total + i
    
print("the final result is:", total)

the number is: 0
the number is: 1
the number is: 2
the number is: 3
the number is: 4
the final result is: 10


## your output looks like a pyramid or shows a correct but incremental version of the final output

This happens when your final result statement is inside your for loop, causing it to be executed each loop.  The output this produces a characteristic pyramid effect.

In [3]:
s = ""

for i in range(1000, 1010):
    s = s + chr(i) # don't worry about what's going on in here yet
    print("the final string is:", s)

the final string is: Ϩ
the final string is: Ϩϩ
the final string is: ϨϩϪ
the final string is: ϨϩϪϫ
the final string is: ϨϩϪϫϬ
the final string is: ϨϩϪϫϬϭ
the final string is: ϨϩϪϫϬϭϮ
the final string is: ϨϩϪϫϬϭϮϯ
the final string is: ϨϩϪϫϬϭϮϯϰ
the final string is: ϨϩϪϫϬϭϮϯϰϱ


See how that result is growing?  The print statement is inside the for loop.  So we're seeing the correct result, but we're watching it grow.  This sort of thing may also cause odd errors when you are dealing with creating other kinds of data structures, where you are adding your incremental results to the full data stucture rather than collecting up a sub result and then adding that.  You'll likely still see a pyramid effect in there somewhere.  The solution is to move the print statement out of the for loop.

In [4]:
s = ""

for i in range(1000, 1010):
    s = s + chr(i)

print("the final string is:", s) # did you see me move to the left?

the final string is: ϨϩϪϫϬϭϮϯϰϱ


In [5]:
raven = """Once upon a midnight dreary, while I pondered, weak and weary,
Over many a quaint and curious volume of forgotten lore—
    While I nodded, nearly napping, suddenly there came a tapping,
As of some one gently rapping, rapping at my chamber door.
“’Tis some visitor,” I muttered, “tapping at my chamber door—
            Only this and nothing more.”

    Ah, distinctly I remember it was in the bleak December;
And each separate dying ember wrought its ghost upon the floor.
    Eagerly I wished the morrow;—vainly I had sought to borrow
    From my books surcease of sorrow—sorrow for the lost Lenore—
For the rare and radiant maiden whom the angels name Lenore—
            Nameless here for evermore.

    And the silken, sad, uncertain rustling of each purple curtain
Thrilled me—filled me with fantastic terrors never felt before;
    So that now, to still the beating of my heart, I stood repeating
    “’Tis some visitor entreating entrance at my chamber door—
Some late visitor entreating entrance at my chamber door;—
            This it is and nothing more.”

    Presently my soul grew stronger; hesitating then no longer,
“Sir,” said I, “or Madam, truly your forgiveness I implore;
    But the fact is I was napping, and so gently you came rapping,
    And so faintly you came tapping, tapping at my chamber door,
That I scarce was sure I heard you”—here I opened wide the door;—
            Darkness there and nothing more.

    Deep into that darkness peering, long I stood there wondering, fearing,
Doubting, dreaming dreams no mortal ever dared to dream before;
    But the silence was unbroken, and the stillness gave no token,
    And the only word there spoken was the whispered word, “Lenore?”
This I whispered, and an echo murmured back the word, “Lenore!”—
            Merely this and nothing more.

    Back into the chamber turning, all my soul within me burning,
Soon again I heard a tapping somewhat louder than before.
    “Surely,” said I, “surely that is something at my window lattice;
      Let me see, then, what thereat is, and this mystery explore—
Let my heart be still a moment and this mystery explore;—
            ’Tis the wind and nothing more!”

    Open here I flung the shutter, when, with many a flirt and flutter,
In there stepped a stately Raven of the saintly days of yore;
    Not the least obeisance made he; not a minute stopped or stayed he;
    But, with mien of lord or lady, perched above my chamber door—
Perched upon a bust of Pallas just above my chamber door—
            Perched, and sat, and nothing more.

Then this ebony bird beguiling my sad fancy into smiling,
By the grave and stern decorum of the countenance it wore,
“Though thy crest be shorn and shaven, thou,” I said, “art sure no craven,
Ghastly grim and ancient Raven wandering from the Nightly shore—
Tell me what thy lordly name is on the Night’s Plutonian shore!”
            Quoth the Raven “Nevermore.”

    Much I marvelled this ungainly fowl to hear discourse so plainly,
Though its answer little meaning—little relevancy bore;
    For we cannot help agreeing that no living human being
    Ever yet was blessed with seeing bird above his chamber door—
Bird or beast upon the sculptured bust above his chamber door,
            With such name as “Nevermore.”

    But the Raven, sitting lonely on the placid bust, spoke only
That one word, as if his soul in that one word he did outpour.
    Nothing farther then he uttered—not a feather then he fluttered—
    Till I scarcely more than muttered “Other friends have flown before—
On the morrow he will leave me, as my Hopes have flown before.”
            Then the bird said “Nevermore.”

    Startled at the stillness broken by reply so aptly spoken,
“Doubtless,” said I, “what it utters is its only stock and store
    Caught from some unhappy master whom unmerciful Disaster
    Followed fast and followed faster till his songs one burden bore—
Till the dirges of his Hope that melancholy burden bore
            Of ‘Never—nevermore’.”

    But the Raven still beguiling all my fancy into smiling,
Straight I wheeled a cushioned seat in front of bird, and bust and door;
    Then, upon the velvet sinking, I betook myself to linking
    Fancy unto fancy, thinking what this ominous bird of yore—
What this grim, ungainly, ghastly, gaunt, and ominous bird of yore
            Meant in croaking “Nevermore.”

    This I sat engaged in guessing, but no syllable expressing
To the fowl whose fiery eyes now burned into my bosom’s core;
    This and more I sat divining, with my head at ease reclining
    On the cushion’s velvet lining that the lamp-light gloated o’er,
But whose velvet-violet lining with the lamp-light gloating o’er,
            She shall press, ah, nevermore!

    Then, methought, the air grew denser, perfumed from an unseen censer
Swung by Seraphim whose foot-falls tinkled on the tufted floor.
    “Wretch,” I cried, “thy God hath lent thee—by these angels he hath sent thee
    Respite—respite and nepenthe from thy memories of Lenore;
Quaff, oh quaff this kind nepenthe and forget this lost Lenore!”
            Quoth the Raven “Nevermore.”

    “Prophet!” said I, “thing of evil!—prophet still, if bird or devil!—
Whether Tempter sent, or whether tempest tossed thee here ashore,
    Desolate yet all undaunted, on this desert land enchanted—
    On this home by Horror haunted—tell me truly, I implore—
Is there—is there balm in Gilead?—tell me—tell me, I implore!”
            Quoth the Raven “Nevermore.”

    “Prophet!” said I, “thing of evil!—prophet still, if bird or devil!
By that Heaven that bends above us—by that God we both adore—
    Tell this soul with sorrow laden if, within the distant Aidenn,
    It shall clasp a sainted maiden whom the angels name Lenore—
Clasp a rare and radiant maiden whom the angels name Lenore.”
            Quoth the Raven “Nevermore.”

    “Be that word our sign of parting, bird or fiend!” I shrieked, upstarting—
“Get thee back into the tempest and the Night’s Plutonian shore!
    Leave no black plume as a token of that lie thy soul hath spoken!
    Leave my loneliness unbroken!—quit the bust above my door!
Take thy beak from out my heart, and take thy form from off my door!”
            Quoth the Raven “Nevermore.”

    And the Raven, never flitting, still is sitting, still is sitting
On the pallid bust of Pallas just above my chamber door;
    And his eyes have all the seeming of a demon’s that is dreaming,
    And the lamp-light o’er him streaming throws his shadow on the floor;
And my soul from out that shadow that lies floating on the floor
            Shall be lifted—nevermore!"""

# Problem statement:

Write a program that prints out just the first line of each stanza in the raven, but these lines should be collected into a single string (a common need when writing results out to a file).

Two choices here:

1. Do something fancy and detect the true meaning of what it means to be the first line of a stanza.
2. Exploit the known structure of the poem and do a little math about where each stanza should be.

For now, let's start with the second option and see if this exploit will do the trick for us.  Hopefully it will and we can avoid some complication about the true meaning of a stanza.

## First question:  what is the granularity of our question?

For example, are we concerned about characters, words, lines, larger chunks, entire files, etc?  Answering this question will help you slim down your choices of which patterns, functions, and tools will be helpful. This is also how you should internally organize these things in your head.

For example, you have already learned that there are ways to slice out chunks of strings if characters are the granularity of interest.

You could say that the individual stanza is our granularity of interest.  Which isn't untrue, but python doesn't really have a sense of the meaning of stanzas.  This means that we would need to figure out how to define a stanza from a technical perspective.  Which we can, but isn't entirely necessary.  We'll explore how to do this in later weeks.

Taking another look at our problem statement, we can reconsider what we really want out of this thing.  Which is a bunch of individual lines.  We are going to encode our understanding of stanzas in a later step.

The fact that we care about lines of the poem and that the poem is already stored as a series of lines makes processing it so much easier.

So, we care about lines.  This opens us up to some nice processing options because lines of data are so normal that there are some great tools to work with them.

On a technical basis, we can use the `my_string.split('\n')` to split a string with newlines on the newline characters.  This will create a list where each line is an individual line, and thus we can loop over the individual lines.

You'll be using `.split()` in a lot of your programs, so get comfortable with it!


## Second question:  how can we determine which lines that we want?

I'll go ahead and give you a spoiler:  we can easily look up the line of text that we want if we know which line number that we want.

So given that, we need to sort out how to correctly predict where those lines are.

Given that the structure of the poem is so steady (18 stanzas with 6 lines each), we should be able to reasonably construct a little formula for where the first line of every stanza is.

For example, if we say that I'm going to start with my left little finger and touch every other finger you can likely sort out which ones that will be without having to watch me in action.

You know the following things:

1. I have 10 fingers.
2. This is couting, so there will be only integer numbers.
3. 'Every other' is counting by two.

Let's remind ourselves of the basic use of range:

1. `range(exclusive_stop)`
2. `range(inclusive_start, exclusive_stop)`
3. `range(inclusive_start, exclusive_stop, step)`

The definition of these things are the same as what you learned for string slicing.

Let's count like python and say that my left hand has fingers 0-4, and my right hand has fingers 5-9.  Since I'm starting at 0 and need ten numbers after and including 0, which form can I use?

Normally we'd say that form 1 would be fine, because `range(10)` will give us 0-9 inclusive.

In [6]:
print(list(range(10)))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


But remember that we need to go by every other.  So we need to specify a step of 2.  The only form that allows us to specify a step is form 3, which also requires that we spefify a start.

In [7]:
print(list(range(0, 10, 2)))

[0, 2, 4, 6, 8]


So you can predict that I'll point to my left little finger, my left middle finger, left thumb, right pointer, and right ring.  You can correctly make this prediction because my hands have an orderly dependable structure and I need a steady pattern.  This kind of system will also work on most (not all!) humans.

Our treatment of the stanzas will be no different.  These mechanisms that exploit the dependability of data file can be quite elegant, but come with risk:  you need to be very certain that your data is actually dependable.  

So here we're going to talk a lot about non coding things, because we shouldn't forget:  the code is just our tool.  We need to be able to think of these solutions and approaches that we can use code for, but the coding comes second to all that.

We can explore how to check that dependability.  Let's pop up to our poem text and look at it from a technical perspective, then try to come up with some checks to see if our file is on track.

Things we know (as in, the things that will inform us about what we need to do and how we might handle things):

* there's nothing before or after the poem, as in:  the first line of the poem is the first line, and the last line is the last line.  So nothing to snip out before or after.
* each stanza is 6 lines followed by a single empty line, except for the last stanza.

Can we build up a rudamentary method of calculaing how many lines there will be is our list of lines?  Yes we can!  We're going to define some known variables.

So if stanza length is 6:

In [8]:
stanza_len = 6

And there are 18 stanzas:

In [9]:
num_stanzas = 18

The number of buffer lines between the individual stanzas is 1 less than the total number of stanzas (because remember that there are no trailing empty lines after the poem, but this will not always be the case):

In [10]:
num_blanks = num_stanzas - 1

So we put them all together (with some multiplicaton):

In [11]:
expected_poem_len = (stanza_len * num_stanzas) + num_blanks
print(expected_poem_len)

125


Here we've used python as a calculator, where we encoded our core formula for length, gave it some values, and then made the final calculation.  

What we need to do is check what the actual length is.  We first need to do the split on the poem string and then we can run it through `len()`.

In [12]:
raven_lines = raven.split('\n')

print(len(raven_lines))

125


Excellent!  Our prediction based off observations in the data worked.

What would a result of 6,646 have meant?

In [13]:
print(len(raven)) # that we were calculating the length of the entire string rather than the number of lines.

6646


# List slicing basics

Let's take a break here and learn a little about list slicing.  The basic syntax is the same as what you learned about how to slice strings.  Except that it is working at the level of element (line in this case, because that's what we built our list with when we used `.split()`) rather than character.

I can give it a single postion number and get back the corresponding lines.

In [14]:
print(raven_lines[2])

    While I nodded, nearly napping, suddenly there came a tapping,


So let's get this range going.  We can slice out the beginning of our list to give ourselves something to base an initial try at.  This is the power of using lists to sort our data.  When we can split it apart such that the position numbers can have meaning, either directly via their numbers or indirectly but referencable), we can make our code clean and meaningful.  

In [15]:
print(raven_lines[:(stanza_len * 3) + 3])

['Once upon a midnight dreary, while I pondered, weak and weary,', 'Over many a quaint and curious volume of forgotten lore—', '    While I nodded, nearly napping, suddenly there came a tapping,', 'As of some one gently rapping, rapping at my chamber door.', '“’Tis some visitor,” I muttered, “tapping at my chamber door—', '            Only this and nothing more.”', '', '    Ah, distinctly I remember it was in the bleak December;', 'And each separate dying ember wrought its ghost upon the floor.', '    Eagerly I wished the morrow;—vainly I had sought to borrow', '    From my books surcease of sorrow—sorrow for the lost Lenore—', 'For the rare and radiant maiden whom the angels name Lenore—', '            Nameless here for evermore.', '', '    And the silken, sad, uncertain rustling of each purple curtain', 'Thrilled me—filled me with fantastic terrors never felt before;', '    So that now, to still the beating of my heart, I stood repeating', '    “’Tis some visitor entreating entrance 

This is not fun to read, let's put it into a for loop so everything prints out left jusified, and we're going to use a counter for funzies.

In [16]:
line_num = 0

for line in raven_lines[:(stanza_len * 3) + 3]:
    print(line_num, ":", line)
    line_num += 1 # a new notation!  equivalent to line_num = line_num + 1

0 : Once upon a midnight dreary, while I pondered, weak and weary,
1 : Over many a quaint and curious volume of forgotten lore—
2 :     While I nodded, nearly napping, suddenly there came a tapping,
3 : As of some one gently rapping, rapping at my chamber door.
4 : “’Tis some visitor,” I muttered, “tapping at my chamber door—
5 :             Only this and nothing more.”
6 : 
7 :     Ah, distinctly I remember it was in the bleak December;
8 : And each separate dying ember wrought its ghost upon the floor.
9 :     Eagerly I wished the morrow;—vainly I had sought to borrow
10 :     From my books surcease of sorrow—sorrow for the lost Lenore—
11 : For the rare and radiant maiden whom the angels name Lenore—
12 :             Nameless here for evermore.
13 : 
14 :     And the silken, sad, uncertain rustling of each purple curtain
15 : Thrilled me—filled me with fantastic terrors never felt before;
16 :     So that now, to still the beating of my heart, I stood repeating
17 :     “’Tis some v

Cool cool, now we have our lines with the numbers here for reference.

Looks like we would want line 0, 7, and 14.  So maybe a step of 7 will do the trick?  We can try it.

In [17]:
print(list(range(0, len(raven_lines), 7)))

[0, 7, 14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84, 91, 98, 105, 112, 119]


So far so good.  Looks like the last one is 119.  Can we do a slicing trick to see the last few lines of the poem?

We'll start a few stanzas before the end of the poem. The place to start is given by the last line number less the number of lines that we want to show. We could also have used a negative step, but that can be harder to wrap your head around.

In [18]:
num_of_lines = ((stanza_len * 3) + 3) # we'll want the last three stanzas, plus a few empty lines
starting_line_num = len(raven_lines) - num_of_lines # subtract the number of lines we want from the total length

for line in raven_lines[starting_line_num:]: # how convenient we've already done the starting calculation
    print(line_num, ":", line)
    line_num += 1

104 : 
105 :     “Prophet!” said I, “thing of evil!—prophet still, if bird or devil!
106 : By that Heaven that bends above us—by that God we both adore—
107 :     Tell this soul with sorrow laden if, within the distant Aidenn,
108 :     It shall clasp a sainted maiden whom the angels name Lenore—
109 : Clasp a rare and radiant maiden whom the angels name Lenore.”
110 :             Quoth the Raven “Nevermore.”
111 : 
112 :     “Be that word our sign of parting, bird or fiend!” I shrieked, upstarting—
113 : “Get thee back into the tempest and the Night’s Plutonian shore!
114 :     Leave no black plume as a token of that lie thy soul hath spoken!
115 :     Leave my loneliness unbroken!—quit the bust above my door!
116 : Take thy beak from out my heart, and take thy form from off my door!”
117 :             Quoth the Raven “Nevermore.”
118 : 
119 :     And the Raven, never flitting, still is sitting, still is sitting
120 : On the pallid bust of Pallas just above my chamber door;
121 :     

So based on this, we want the last few items to be lines 105, 112, and 119.  Which we have in our list of items above!  So let's just do this.

We've got our range that produces the line numbers that we want, so we can loop over those numbers and use them (one at a time) to look up the lines that we want.

In [19]:
for i in range(0, len(raven_lines), 7):
    print(i, ":", raven_lines[i])

0 : Once upon a midnight dreary, while I pondered, weak and weary,
7 :     Ah, distinctly I remember it was in the bleak December;
14 :     And the silken, sad, uncertain rustling of each purple curtain
21 :     Presently my soul grew stronger; hesitating then no longer,
28 :     Deep into that darkness peering, long I stood there wondering, fearing,
35 :     Back into the chamber turning, all my soul within me burning,
42 :     Open here I flung the shutter, when, with many a flirt and flutter,
49 : Then this ebony bird beguiling my sad fancy into smiling,
56 :     Much I marvelled this ungainly fowl to hear discourse so plainly,
63 :     But the Raven, sitting lonely on the placid bust, spoke only
70 :     Startled at the stillness broken by reply so aptly spoken,
77 :     But the Raven still beguiling all my fancy into smiling,
84 :     This I sat engaged in guessing, but no syllable expressing
91 :     Then, methought, the air grew denser, perfumed from an unseen censer
98 :     “P

We can see that this has worked, and also another pattern:  most of these start with an indent, but not all.  Going back into the poem, we can see that even though lines 0 and 49 do not have these indents, but they are indeed the first lines.  So we are correct in our computaion of where things are even though the poem has errors.  This is a good case of data having the appearance of regularity but it being inconsistent.  I'm glad we didn't depend on that!

One interesting thing to note is that we could have skipped the for loop entirely if we only wanted the line text out.  Using our slicing skills we can more directly ask for every 7th item in the list.  The advantage of using range is that we have the line numbers to use for referece or reporting.

In [20]:
raven_lines[::7]

['Once upon a midnight dreary, while I pondered, weak and weary,',
 '    Ah, distinctly I remember it was in the bleak December;',
 '    And the silken, sad, uncertain rustling of each purple curtain',
 '    Presently my soul grew stronger; hesitating then no longer,',
 '    Deep into that darkness peering, long I stood there wondering, fearing,',
 '    Back into the chamber turning, all my soul within me burning,',
 '    Open here I flung the shutter, when, with many a flirt and flutter,',
 'Then this ebony bird beguiling my sad fancy into smiling,',
 '    Much I marvelled this ungainly fowl to hear discourse so plainly,',
 '    But the Raven, sitting lonely on the placid bust, spoke only',
 '    Startled at the stillness broken by reply so aptly spoken,',
 '    But the Raven still beguiling all my fancy into smiling,',
 '    This I sat engaged in guessing, but no syllable expressing',
 '    Then, methought, the air grew denser, perfumed from an unseen censer',
 '    “Prophet!” said I

But going back to our problem statement, we can see that we're not all done.  We need to collect all the lines into a single string.  Right now we have a way to get all the lines we want, we just need to figure out how to get them all into a single string.

Let's look at a few ways to do this.

* What we have:  a bunch of strings
* What we need:  a single string
* What we need to decide:  what content should go between them.

When we need to collect up a string we need to decide what our final form looks like.  Do we want it to look like the original?  So each 'line' as an actual line.  Connected like something else?

Chances are you can presume pretty correctly that when we break something apart,  we reconnect the pieces we want  the reconstruction to look like the original.  

So what is the meaning of a 'line'?

* Technically:  some not newline stuff (including nothing) followed by a newline character (\n)
* Humanly:  normal looking text, likely on top of or below some other normal looking text

So if we want to reconstruct these pieces of text back into something that looks like a poem, we need to consider those newline characters.  Sometimes they are retained in the lines when we break things apart, but sometimes they are lost.

In [21]:
raven_lines[::7]

['Once upon a midnight dreary, while I pondered, weak and weary,',
 '    Ah, distinctly I remember it was in the bleak December;',
 '    And the silken, sad, uncertain rustling of each purple curtain',
 '    Presently my soul grew stronger; hesitating then no longer,',
 '    Deep into that darkness peering, long I stood there wondering, fearing,',
 '    Back into the chamber turning, all my soul within me burning,',
 '    Open here I flung the shutter, when, with many a flirt and flutter,',
 'Then this ebony bird beguiling my sad fancy into smiling,',
 '    Much I marvelled this ungainly fowl to hear discourse so plainly,',
 '    But the Raven, sitting lonely on the placid bust, spoke only',
 '    Startled at the stillness broken by reply so aptly spoken,',
 '    But the Raven still beguiling all my fancy into smiling,',
 '    This I sat engaged in guessing, but no syllable expressing',
 '    Then, methought, the air grew denser, perfumed from an unseen censer',
 '    “Prophet!” said I

Taking another look here, we don't see any telltale `'\n'` at the end of the strings nor are extra lines being printed out, so we no longer have our newlines in our text.  They were removed when we split the original string apart with `.split()`.  This method is what we call 'consumptive', meaning that it consumes (gets rid of) whatever string you pass it.  

So `.split()` removed our newline character, and now we need to add it back in at the end of each line.  We want it to look like a poem again.

Your appoach for how to do this depends on what all you need to do with the content.  

* Just the newline?  You can use a method like `.join()`
* More stuff or need to further process the lines?  Do an accumulator loop.

Let's demo the easy one first, with `.join()`.

## `.join()`

The syntax for this method is one of the oddest that you'll find, and I still find it a little boggling to this day.  Anyhow, homework 1 required you to make it work, even if you didn't quite understand it.  So you've seen this before, but here's a reminder of the syntax:

Essential definitons:

* `t` is the string (either the literal string or a variable) that you want to be in the middle of things.
* `sequence` in the stuff that you want to join together.  This will most likely be a list, but could be something else like a string. Any sequence that you could use a for loop on will work here, and the pieces that it picks out of the sequence will be what a for loop can access.  This caveat will make sense if you try to use other things like a dictionary in here, but most of the time you wont and don't need to worry.

`t.join(sequence)`

Yes, you are reading that right.  You call the method on the string that you want to join the sequence with.  I understand and agree with the technical reasons for this, but that has never stopped me from always giving this a side eye.

Let's see it in action.

* What do we want to join? `raven_lines[::7]`  (this is our sequence)
* What do we want to appear between each line?  A `\n` character.  (this is our t)

In [22]:
print("\n".join(raven_lines[::7]))

Once upon a midnight dreary, while I pondered, weak and weary,
    Ah, distinctly I remember it was in the bleak December;
    And the silken, sad, uncertain rustling of each purple curtain
    Presently my soul grew stronger; hesitating then no longer,
    Deep into that darkness peering, long I stood there wondering, fearing,
    Back into the chamber turning, all my soul within me burning,
    Open here I flung the shutter, when, with many a flirt and flutter,
Then this ebony bird beguiling my sad fancy into smiling,
    Much I marvelled this ungainly fowl to hear discourse so plainly,
    But the Raven, sitting lonely on the placid bust, spoke only
    Startled at the stillness broken by reply so aptly spoken,
    But the Raven still beguiling all my fancy into smiling,
    This I sat engaged in guessing, but no syllable expressing
    Then, methought, the air grew denser, perfumed from an unseen censer
    “Prophet!” said I, “thing of evil!—prophet still, if bird or devil!—
    “P

And if we had gotten it the wrong way around?  We get an error, becuase we are trying to call the join method on a data type that doesn't have it.  So while you can pass join (remember that passing is where you put stuff in the `()`) any sequence, you can only call `.join()` on a string.  Meaning that the thing with the `.join()` directly after must be a string (literal of variable).

In [23]:
print(raven_lines[::7].join('\n'))

AttributeError: 'list' object has no attribute 'join'

So that was the easy way, but you can sort of see where the limitation is.  Our line content still has that weird formatting issue where most, but not all, of the lines have the indent.  Without direct access to the content in this process, we're kind of stuck with that spacing.  So we would have needed to do the cleanup earlier in our process (sometimes we've already been in there messing with stuff before we do our filtering, so this isn't a concern in the final prep preocess) but other times we haven't needed to touch the content before this point (which is our current situation).  This is a place where your personal style will come into play a lot.  You may have already dealt with this, or had a better opportunity to deal with this, depending on what you might have needed to do in your program before this point.  In this case, we haven't needed to do much else and just need to clean up a few things, so combining it in with our final collection process makes sense.  There will be other time when you'll want to keep everything separate.

Let's do this again with an accumulator to have more direct control.

## Accumulator

Following with our usual pattern, we need to have:

* an empty string to collect stuff into
* a for loop over our desired lines
* a cleaned up line with a newline to add to our collection (which we'll make during our loop)

In [24]:
clean_lines = "" # our empty collection string

for line in raven_lines[::7]:
    cleaned_line = line.strip() # this will get rid of the indent
    cleaned_line += '\n' # now we add the newline
    clean_lines += cleaned_line # now add the clean line to the end of the collection

Now we can see what we've done!

In [25]:
print(clean_lines)

Once upon a midnight dreary, while I pondered, weak and weary,
Ah, distinctly I remember it was in the bleak December;
And the silken, sad, uncertain rustling of each purple curtain
Presently my soul grew stronger; hesitating then no longer,
Deep into that darkness peering, long I stood there wondering, fearing,
Back into the chamber turning, all my soul within me burning,
Open here I flung the shutter, when, with many a flirt and flutter,
Then this ebony bird beguiling my sad fancy into smiling,
Much I marvelled this ungainly fowl to hear discourse so plainly,
But the Raven, sitting lonely on the placid bust, spoke only
Startled at the stillness broken by reply so aptly spoken,
But the Raven still beguiling all my fancy into smiling,
This I sat engaged in guessing, but no syllable expressing
Then, methought, the air grew denser, perfumed from an unseen censer
“Prophet!” said I, “thing of evil!—prophet still, if bird or devil!—
“Prophet!” said I, “thing of evil!—prophet still, if bird 

Now that we have this pattern, we can play with changing other things, such as adding a line number.

In [26]:
clean_lines = "" # our empty collection string

counter = 1 # starting at 1 because we're human counting

for line in raven_lines[::7]:
    cleaned_line = "Stanza " + str(counter).zfill(2) + ": "
    cleaned_line += line.strip() # note this is += now
    cleaned_line += '\n' # this can stay as +=
    clean_lines += cleaned_line # now add the clean line to the end of the collection
    counter +=1 # increment the counter
    
print(clean_lines)

Stanza 01: Once upon a midnight dreary, while I pondered, weak and weary,
Stanza 02: Ah, distinctly I remember it was in the bleak December;
Stanza 03: And the silken, sad, uncertain rustling of each purple curtain
Stanza 04: Presently my soul grew stronger; hesitating then no longer,
Stanza 05: Deep into that darkness peering, long I stood there wondering, fearing,
Stanza 06: Back into the chamber turning, all my soul within me burning,
Stanza 07: Open here I flung the shutter, when, with many a flirt and flutter,
Stanza 08: Then this ebony bird beguiling my sad fancy into smiling,
Stanza 09: Much I marvelled this ungainly fowl to hear discourse so plainly,
Stanza 10: But the Raven, sitting lonely on the placid bust, spoke only
Stanza 11: Startled at the stillness broken by reply so aptly spoken,
Stanza 12: But the Raven still beguiling all my fancy into smiling,
Stanza 13: This I sat engaged in guessing, but no syllable expressing
Stanza 14: Then, methought, the air grew denser, perf

In this final form you can see that we're using both a counter accumulator and a collection accumulator.  We haven't gone back to our original use of range here because we haven't needed to use the line numbers.  In the previous line we only printed out the staza number, so using a simple counter (or enumerate) along with the string slicing have done the trick just fine.  I would need to add my range back in if I wanted that line number, but I could leave my counter in so I would have both the line and the stanza.  

Can you experiment with making that work?

In [28]:
clean_lines = "" # our empty collection string

counter = 1 # starting at 1 becuase we're human counting

for i in range(0, len(raven_lines), 7):
    cleaned_line = "Stanza " + str(counter).zfill(2) + " line " + str(i).zfill(3) + ": " #adding line num
    cleaned_line += raven_lines[i].strip() # add the lookup here
    cleaned_line += '\n' # this can stay as +=
    clean_lines += cleaned_line # now add the clean line to the end of the collection
    counter +=1 # increment the counter
    
print(clean_lines)

Stanza 01 line 000: Once upon a midnight dreary, while I pondered, weak and weary,
Stanza 02 line 007: Ah, distinctly I remember it was in the bleak December;
Stanza 03 line 014: And the silken, sad, uncertain rustling of each purple curtain
Stanza 04 line 021: Presently my soul grew stronger; hesitating then no longer,
Stanza 05 line 028: Deep into that darkness peering, long I stood there wondering, fearing,
Stanza 06 line 035: Back into the chamber turning, all my soul within me burning,
Stanza 07 line 042: Open here I flung the shutter, when, with many a flirt and flutter,
Stanza 08 line 049: Then this ebony bird beguiling my sad fancy into smiling,
Stanza 09 line 056: Much I marvelled this ungainly fowl to hear discourse so plainly,
Stanza 10 line 063: But the Raven, sitting lonely on the placid bust, spoke only
Stanza 11 line 070: Startled at the stillness broken by reply so aptly spoken,
Stanza 12 line 077: But the Raven still beguiling all my fancy into smiling,
Stanza 13 line 

Let's adapt our current program to take user input.  You prompt the user to ask which line of each stanza they would like.

What's our first step?  Before we start mucking about with our current program, we need to think about how we can change our start, stop, step to accomplish this.  We can think back to our example about the hands.  Suppose that I lose my left pinky finger, but you give me the same rules and I abide by the same number convention despite lacking one of my fingers.  How would you change our original range statement? The only change that we have in play is a change in our start position.

So a few facts:

* no matter which line in each stanza is requested, the consistent nature of the poem means that it'll always be every 7 lines.
    * so our step value of 7 can remain the same
* no matter which line in each stanza is requested, we want to always go to the end.
    * so we can leave our stop value as unspecified
* the offset number should be between 0 and 6 (0-5 are the 6 lines and offset 6 is the blank line)
    
We need to go consistently every 7 places, but starting at a different point.  So our start position can be whatever the offset is.

Actually, that's all we need to do.  Once we shift the starting point, all the others will fall in the right place.  Let's test this first.

(we're going to use the word offset to mean the number of spaces that are off our normal position)

In [29]:
offset = 1

clean_lines = "" # our empty collection string

counter = 1 # starting at 1 becuase we're human counting

for line in raven_lines[offset::7]:
    cleaned_line = "Stanza " + str(counter).zfill(2) + ": "
    cleaned_line += line.strip() # note this is += now
    cleaned_line += '\n' # this can stay as +=
    clean_lines += cleaned_line # now add the clean line to the end of the collection
    counter +=1 # increment the counter
    
print(clean_lines)

Stanza 01: Over many a quaint and curious volume of forgotten lore—
Stanza 02: And each separate dying ember wrought its ghost upon the floor.
Stanza 03: Thrilled me—filled me with fantastic terrors never felt before;
Stanza 04: “Sir,” said I, “or Madam, truly your forgiveness I implore;
Stanza 05: Doubting, dreaming dreams no mortal ever dared to dream before;
Stanza 06: Soon again I heard a tapping somewhat louder than before.
Stanza 07: In there stepped a stately Raven of the saintly days of yore;
Stanza 08: By the grave and stern decorum of the countenance it wore,
Stanza 09: Though its answer little meaning—little relevancy bore;
Stanza 10: That one word, as if his soul in that one word he did outpour.
Stanza 11: “Doubtless,” said I, “what it utters is its only stock and store
Stanza 12: Straight I wheeled a cushioned seat in front of bird, and bust and door;
Stanza 13: To the fowl whose fiery eyes now burned into my bosom’s core;
Stanza 14: Swung by Seraphim whose foot-falls tink

And now we can add in an input:

In [30]:
offset = eval(input("Which line of each stanza do you want?:"))

clean_lines = "" # our empty collection string

counter = 1 # starting at 1 becuase we're human counting

for line in raven_lines[offset::7]:
    cleaned_line = "Stanza " + str(counter).zfill(2) + ": "
    cleaned_line += line.strip() # note this is += now
    cleaned_line += '\n' # this can stay as +=
    clean_lines += cleaned_line # now add the clean line to the end of the collection
    counter +=1 # increment the counter
    
print(clean_lines)

Which line of each stanza do you want?:3
Stanza 01: As of some one gently rapping, rapping at my chamber door.
Stanza 02: From my books surcease of sorrow—sorrow for the lost Lenore—
Stanza 03: “’Tis some visitor entreating entrance at my chamber door—
Stanza 04: And so faintly you came tapping, tapping at my chamber door,
Stanza 05: And the only word there spoken was the whispered word, “Lenore?”
Stanza 06: Let me see, then, what thereat is, and this mystery explore—
Stanza 07: But, with mien of lord or lady, perched above my chamber door—
Stanza 08: Ghastly grim and ancient Raven wandering from the Nightly shore—
Stanza 09: Ever yet was blessed with seeing bird above his chamber door—
Stanza 10: Till I scarcely more than muttered “Other friends have flown before—
Stanza 11: Followed fast and followed faster till his songs one burden bore—
Stanza 12: Fancy unto fancy, thinking what this ominous bird of yore—
Stanza 13: On the cushion’s velvet lining that the lamp-light gloated o’er,
S